### PR TITLE
Reference related cleanup/refactoring

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Follow up to https://github.com/crate/crate/pull/14824

Addressing comments 

https://github.com/crate/crate/pull/14824#discussion_r1351867133
https://github.com/crate/crate/pull/14824#discussion_r1351865300
https://github.com/crate/crate/pull/14824#discussion_r1351862856

### CODE REVIEW

File: server/src/main/java/io/crate/metadata/IndexReference.java [hunk 1]
```diff
@@ -273,6 +258,40 @@ public Reference withColumnOid(LongSupplier oidSupplier) {
         );
     }
 
+    @Override
+    public Reference withDropped(boolean dropped) {
+        return new IndexReference(
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/metadata/IndexReference.java:263]: Just grouped with... methods together

File: server/src/main/java/io/crate/metadata/IndexReference.java [hunk 2]
```diff
@@ -237,22 +237,7 @@ public Reference getRelocated(ReferenceIdent newIdent) {
...
-        }
-        mapping.put("type", "text");
-
-        if (columns.isEmpty() == false) {
-            mapping.put("sources", columns.stream().map(Reference::storageIdent).toList());
-        }
-
-        return mapping;
-    }
-
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/metadata/IndexReference.java:253]: Just grouped `with...` methods together

File: server/src/main/java/io/crate/execution/ddl/tables/AddColumnTask.java [hunk 3]
```diff
@@ -163,8 +164,9 @@ static List<Reference> normalizeColumns(AddColumnRequest request, DocTableInfo c
 
         // Add missing parents
         for (var newColumn : request.references()) {
-            ColumnIdent parent = newColumn.column();
-            while ((parent = parent.getParent()) != null) {
+            Iterator<ColumnIdent> iterator = newColumn.column().parents().iterator();
+            while (iterator.hasNext()) {
+                ColumnIdent parent = iterator.next();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/ddl/tables/AddColumnTask.java:169]:
```suggestion
            for (ColumnIdent parent : newColumn.column().parents()) {
```

File: server/src/main/java/io/crate/execution/ddl/tables/DropColumnTask.java [hunk 4]
```diff
@@ -81,19 +82,16 @@ public ClusterState execute(ClusterState currentState, DropColumnRequest request
...
         // Prepare columns tree, it will be used twice:
         // 1. For creating mapping for dropped columns.
         // 2. For creating "path-only, no properties" mapping to remove.
         List<Reference> references = new ArrayList<>(refsToDrop);
         for (var colToDrop : refsToDrop) {
-            ColumnIdent parent = colToDrop.column();
-            while ((parent = parent.getParent()) != null) {
+            Iterator<ColumnIdent> iterator = colToDrop.column().parents().iterator();
+            while (iterator.hasNext()) {
+                ColumnIdent parent = iterator.next();
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/ddl/tables/DropColumnTask.java:94]:
```suggestion
            for (ColumnIdent parent : colToDrop.column().parents()) {
```

**Review by mfussenegger — APPROVED** [review]
:+1:

**Comment by BaurzhanSakhariev** [comment]
https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/10550/testReport/junit/io.crate.integrationtests/DynamicMappingUpdateITest/test_concurrent_statements_that_add_columns_result_in_dynamic_mapping_updates/

might be related to the first change but `parent.getParent` in a `while` loop looks same internally to using `Iterable` parents.

Added some iterations to check whether flakiness related or not

**Comment by BaurzhanSakhariev** [comment]
> https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/10550/testReport/junit/io.crate.integrationtests/DynamicMappingUpdateITest/test_concurrent_statements_that_add_columns_result_in_dynamic_mapping_updates/
> 
> might be related to the first change but `parent.getParent` in a `while` loop looks same internally to using `Iterable` parents.
> 
> Added some iterations (temporal) to check whether flakiness related or not

So, in this PR this test is flaky. Checking it on master in https://github.com/crate/crate/pull/14853 to check whether it's related to the first commit or not.

**Comment by matriv** [comment]
Can it be related to the fact that the setDropped->withDropped now creates new instances?
I also saw that `isDropped` is not included in `equals()` and `hashCode()`

**Comment by BaurzhanSakhariev** [comment]
> Can it be related to the fact that the setDropped->withDropped now creates new instances?

Don't think so, flaky test doesn't use drop column. Specifically, it's some INSERT statements with sub-column failing to update the mapping.

[Reverting](https://github.com/crate/crate/pull/14850/commits/31eb154ee5a91b717293ccd7bec84006f587d251) `ColumnIdent.parent()` usage in favour of explicit access to parent in `while` loop helps to resolve the flakiness but I think we shouldn't just revert but rather investigate the root cause. 

So far comparing while logic with parent() logic looks similar. One dfference which is not critical: parent() don't have special handling for path Len = 1, it will create `ColumnIdent(name, subList(0,0))` --> `ColumnIdent(name, emptyList) `which is same as `ColumnName(name)` used in direct parent access, which internally also creating creates an empty list.


> I also saw that `isDropped` is not included in `equals()` and `hashCode()`

Thanks, I will add a new commit here to address that

**Comment by BaurzhanSakhariev** [comment]
I reverted https://github.com/crate/crate/commit/659f0321847c268eb5b6e9d0bbc260e392832125 because there was a flakiness. It will be investigated separately in order not to block other changes. 

>I also saw that isDropped is not included in equals() and hashCode()

Thanks @matriv, done in https://github.com/crate/crate/pull/14850/commits/62a9345d82fe7c65bb2144130a3693b479524a7a

**Pull request merged**